Avalonia.Labs.Gif/GifImage.cs

```diff
@@ -0,0 +1,209 @@
+
+    public static readonly StyledProperty<IterationCount> IterationCountProperty =
+        AvaloniaProperty.Register<GifImage, IterationCount>("IterationCount", IterationCount.Infinite);
+
+    private GifInstance? _gifInstance;
+
+    public static readonly StyledProperty<StretchDirection> StretchDirectionProperty =
+        AvaloniaProperty.Register<GifImage, StretchDirection>("StretchDirection");
+
+    public static readonly StyledProperty<Stretch> StretchProperty =
+        AvaloniaProperty.Register<GifImage, Stretch>("Stretch");
+
+    private CompositionCustomVisual? _customVisual;
+
+    private object? _initialSource;
+
+    protected override void OnPropertyChanged(AvaloniaPropertyChangedEventArgs change)
+    {
+        switch (change.Property.Name)
+        {
+            case nameof(SourceUri):
+            case nameof(SourceStream):
+                SourceChanged(change);
+                break;
+            case nameof(Stretch):
+            case nameof(StretchDirection):
+                InvalidateArrange();
+                InvalidateMeasure();
+                Update();
+                break;
+            case nameof(IterationCount):
+                IterationCountChanged(change);
+                break;
+            case nameof(Bounds):
+                Update();
+                break;
+        }
+
+        base.OnPropertyChanged(change);
+    }
+
+    public Uri SourceUri
+    {
+        get => GetValue(SourceUriProperty);
+        set => SetValue(SourceUriProperty, value);
+    }
+
+    public Stream SourceStream
+    {
+        get => GetValue(SourceStreamProperty);
+        set => SetValue(SourceStreamProperty, value);
+    }
+
+    public IterationCount IterationCount
+    {
+        get => GetValue(IterationCountProperty);
+        set => SetValue(IterationCountProperty, value);
+    }
+
+    public StretchDirection StretchDirection
+    {
+        get => GetValue(StretchDirectionProperty);
+        set => SetValue(StretchDirectionProperty, value);
+    }
+
+    public Stretch Stretch
+    {
+        get => GetValue(StretchProperty);
+        set => SetValue(StretchProperty, value);
+    }
+
+    private static void IterationCountChanged(AvaloniaPropertyChangedEventArgs e)
+    {
+        var image = e.Sender as GifImage;
+        if (image is null || e.NewValue is not IterationCount iterationCount)
+            return;
+
+        image.IterationCount = iterationCount;
+    }
+
+    protected override void OnAttachedToVisualTree(VisualTreeAttachmentEventArgs e)
+    {
+        var compositor = ElementComposition.GetElementVisual(this)?.Compositor;
+        if (compositor == null || _customVisual?.Compositor == compositor)
+            return;
+        _customVisual = compositor.CreateCustomVisual(new GifCustomVisualHandler());
+        ElementComposition.SetElementChildVisual(this, _customVisual);
+        _customVisual.SendHandlerMessage(GifCustomVisualHandler.StartMessage);
```

Comment:
You are starting visual handler without any gif-instance set yet. Is it correct?